### PR TITLE
Fix and improve Ident (RFC 1413) code

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
FATAL: assertion failed: FilledChecklist.cc:254: "!rfc931[0]"

The above assertion was caused by ACLFilledChecklist constructor calling
setIdent() twice: via ACLFilledChecklist::setRequest() and directly. The
bug was added in 2021 commit e227da8.

Also fixed a memory leak created by ACLFilledChecklist::verifyAle() that
allocated a never-freed ALE::cache.rfc931 memory if ALE was missing an
available user ID. Configurations with external ACLs leaked a user ID
for each transaction that had one. Squid logged up to 99 "ALE missing
IDENT" messages in those cases. Squid leaked since 2015 commit fbbea662.

These rfc931 field maintenance bugs were caused by duplication of its
storage across three different classes: AccessLogEntry::cache,
Comm::Connection, and Icap::History. Now, this info is only kept in the
corresponding Comm::Connection object which is already accessible from
most contexts via ACLFilledChecklist::conn() and ALE::tcpClient.

ICAP code was missing access to ALE::tcpClient of the originating HTTP
transaction, so we added one more Icap::History field to pass such info.

Squid recognizes or treats specially several Ident-related states:

* no Ident lookup has been performed,
* Ident lookup has failed,
* Ident lookup has succeeded with an ordinary user name, and, arguably,
* Ident lookup succeeded with a user name equal to "-".

To improve our chances of correctly fixing the above bugs and preventing
similar ones, we upgraded Ident-related information storage from a
fixed-size 64-byte array (with magic first character values to signal
Ident states) to two new named types in Ident namespace.

Also improved "ident" and "ident_regex" ACL documentation that did not
disclose many matching quirks and dangerously stated that ident_regex
ACL treats REQUIRED parameter value specially, like "ident" ACL does
(ident_regex has not done that since inception in 2000 commit 145cf928).

Also removed --disable-ident-lookups ./configure option and related
macros. Ident code was already compiled by default, and the option to
disable that code lost its (only) optimization purpose when we removed
hard-coded 64-byte Connection::rfc931 arrays and their copies.

Also improved error-reporting code when parsing user-id in ident-reply,
but problematic parsing code logic remains essentially unchanged. 

Obsoleted Ident RFC 931 uses stale terminology and is less detailed than
its RFC 1413 replacement. For example, the requirement for user-id to
contain at least one octet comes from RFC 1413. Squid has followed that
particular RFC 1413 requirement already. We updated obsoleted RFC
references. Most of the affected text had to be modified anyway.

### CODE REVIEW

File: src/comm/Connection.cc [hunk 1]
```diff
@@ -180,6 +178,21 @@ Comm::Connection::connectTimeout(const time_t fwdStart) const
     return min(ctimeout, ftimeout);
 }
 
+void
+Comm::Connection::updateIdent(const Ident::Lookup &newLookup)
+{
+    // Multiple updateIdent() calls happen when a configuration directive (that
+    // requires a lookup) is evaluated while a lookup attempt is still pending.
+    // Ident::IdentStateData::notify() makes calls with the same lookup value.
```
Comment (kinkie) [inline src/comm/Connection.cc:186]: The comment is not clear - probably a typo but worth fixing

Comment (rousskov) [reply src/comm/Connection.cc:186]: > The comment is not clear - probably a typo but worth fixing

@kinkie, please identify what is not clear or where you see a probable typo. I do not see any particular problems in this comment, so I cannot suggest what to improve.

**Review by rousskov — DISMISSED** [review]
The comments in this review annotate this PR without requesting any PR changes.

File: src/comm/Connection.cc (hunk 1 above)
Comment (rousskov) [reply src/comm/Connection.cc]: @kinkie, thank you for answering my question during the meeting. I rephrased this comment in branch commit e6f9c78ea4d918e1712c401856dcb125de607f21. Please see whether that commit addresses your concern.

BTW, these repeated calls should be eliminated, but that is a separate/old problem that deserves a dedicated PR.

**Comment by yadij** [comment]
Sorry I missed the meeting today where this appears to have been discussed. The absence of a push from @rousskov to drop IDENT code entirely is surprising me, given that other more-commonly-used protocols have been purged from Squid v7.

FTR; IDENT has been considered seriously insecure and broken at the protocol level (i.e. unfixable by design) since at least as far back a 2009 where SANS issued an update recommending its removal from all networks.

Authorization such as IDENT can be performed using an `external_acl_type` helper nowdays, so IMO there is no valid reason to retain IDENT logic within Squid itself.

File: doc/release-notes/release-7.sgml.in [hunk 2]
```diff
@@ -221,6 +221,14 @@ This section gives an account of those changes in three categories:
 	<tag>--enable-cachemgr-hostname=</tag>
 	<p>The <em>cachemgr.cgi</em> tool this option relates to has been removed.
 
+	<tag>--disable-ident-lookups</tag>
+	<p>The primary purpose of keeping Ident lookup feature behind a ./configure
+	option was avoiding performance impact of the corresponding code when the
+	feature was unused. That goal is now achieved without that complication.
```
Comment (yadij) [inline doc/release-notes/release-7.sgml.in:227]: False. "Primary" purpose of all `--disable-foo` options is to allow distributors or admin to prohibit use of components for *any reason they desire*. Nothing more, nothing less.

For this specific `--disable-ident-option`, security is more relevant. Being able to globally lock down Ident Protocol use by end-user configured proxies is a major need for certain types of CDN installation. Firewall rules cannot go far enough to block the security hole they face.

Please restore this build option. It is needed so long as Squid contains code supporting Ident natively.

**Comment by rousskov** [comment]
> FTR; IDENT has been considered seriously insecure and broken at the protocol level (i.e. unfixable by design) since at least as far back a 2009 where SANS issued an update recommending its removal from all networks.
 
> Authorization such as IDENT can be performed using an `external_acl_type` helper nowdays, so IMO there is no valid reason to retain IDENT logic within Squid itself.

Sounds good. Given your recommendation, we will post a PR removing Ident support instead.

File: src/ident/Ident.cc [hunk 3]
```diff
@@ -233,15 +231,25 @@ Ident::ReadReply(const Comm::ConnectionPointer &conn, char *buf, size_t len, Com
...
+    auto ident = strrchr(buf, ':');
+    if (!ident) {
+        state->deleteThis("no colon separator");
+        return;
+    }
+
+    while (xisspace(*++ident));
+
+    if (*ident == '\0') {
+        state->deleteThis("empty user-id");
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/ident/Ident.cc:248]: So the `RFC 931` logic is still being followed despite this PR claims to be upgrading to `RFC 1413` support. :-(

**Review by yadij — CHANGES_REQUESTED** [review]
This PR contains numerous changes claiming that it performs an upgrade from `RFC 931` to `RFC1413` support. However https://github.com/squid-cache/squid/pull/1815/files#r1606895265 indicates that the logic changes necessary for `RFC 1413` compliance have not been included.   Please revert the `RFC1413` comments and documentation text from this PR.

**Comment by yadij** [comment]
> > FTR; IDENT has been considered seriously insecure and broken at the protocol level (i.e. unfixable by design) since at least as far back a 2009 where SANS issued an update recommending its removal from all networks.
> 
> > Authorization such as IDENT can be performed using an `external_acl_type` helper nowdays, so IMO there is no valid reason to retain IDENT logic within Squid itself.
> 
> Sounds good. Given your recommendation, we will post a PR removing Ident support instead.

Okay. Nevermind the late review then. Thank you for removal.

**Comment by eduard-bagdasaryan** [comment]
>  we will post a PR removing Ident support instead

Done in https://github.com/squid-cache/squid/pull/1827.

**Pull request closed without merging**